### DISCUSSION TITLE
Recieving TS2322 typescript error when using react hook form

### DISCUSSION THREAD

**Discussion by aldibatyr** [react-hook-form/react-hook-form#3359] [Q&A]
Hello all, I hope someone would be able to steer me into the right direction. I have a form that uses react hook form and antd components. my code looks like this

import { Button, Form, Input, Radio, Select, Switch } from 'antd'
import React, { useEffect, useState } from 'react'
import { useForm, Controller, SubmitHandler } from 'react-hook-form'
import res from '../res'
import styles from '../styles.css'
import mainStyles from '~/layouts/main/commonStyles.css'
import { LoadingOutlined } from '@ant-design/icons'
import { GeneratePaymentLinkBody, MerchantShop } from '..'
import SelectField from './SelectField'
import config from '~/config'
import { AlertLanguages, PaymentLink } from '~/api/payment-links'
import { MaskedInput } from 'antd-mask-input'
import message from '~/utils/message'
import layoutStyles from '~/layouts/main/styles.css'

interface FormInput {
    shop_id: string
    account_id: string
    invoice_id: string
    amount: number
    language: AlertLanguages
    description: string
    recipient_contact: string
    recipient_contact_sms: string
    notifier_contact: string
    notifier_contact_sms: string
}

function CreateInvoiceForm({ onApply, onClose, showConfirmation }: {
    onApply: (requestBody: GeneratePaymentLinkBody) => Promise<PaymentLink>,
    onClose: () => void,
    showConfirmation: (newPaymentLink: PaymentLink) => void
}) {
    const formItemLayout = {
        labelCol: {
            xs: { span: 28 },
            sm: { span: 8 }
        },
        wrapperCol: {
            xs: { span: 28 },
            sm: { span: 16 }
        }
    }

    const errorStyle = { color: 'red' }

    const { handleSubmit, control, errors, reset } = useForm<FormInput>()
    const [shops, setShops] = useState<MerchantShop[]>([])
    const [notifierSMSSelected, setNotifierSMSSelected] = useState(false)
    const [sendingRequest, setSendingRequest] = useState(false)

    const onSubmit: SubmitHandler<FormInput> = async (data: FormInput) => {
        setSendingRequest(true)
        let requestBody: GeneratePaymentLinkBody
        try {
            const recipientContactSMS = `+7${data.recipient_contact_sms.replace(/\D/g, '')}`
            let notifierContactSMS
            if (data.notifier_contact_sms) {
                notifierContactSMS = `+7${data.notifier_contact_sms.replace(/\D/g, '')}`
            } else {
                notifierContactSMS = ''
            }
            requestBody = {
                ...data,
                notifier_contact_sms: notifierContactSMS,
                recipient_contact_sms: recipientContactSMS,
                expire_period: '3d',
                currency: 'KAZ'
            }
            const newPaymentLink = await onApply(requestBody)
            setSendingRequest(false)
            if (newPaymentLink.hasOwnProperty('code')) {
                throw new Error(newPaymentLink.message)
            }
            showConfirmation(newPaymentLink)
        } catch (error) {
            setSendingRequest(false)
            message.error(error.message)
        }
    }

    const prefillShopsFromApi = async () => {
        const { data } = JSON.parse(localStorage.getItem('auth'))

        const request = await fetch(
            `${config.api.baseUrl}dictionary/shop`,
            {
                method: 'GET',
                headers: new Headers({
                    Authorization: `Bearer ${data.access_token}`
                })
            }
        )

        const shops = await request.json()
        setShops(shops)

        reset({
            shop_id: shops[0].id.toString(),
            account_id: '',
            amount: 100,
            invoice_id: '',
            language: AlertLanguages.kazakh,
            description: '',
            recipient_contact: '',
            recipient_contact_sms: '',
            notifier_contact_sms: '',
            notifier_contact: ''
        })
    }
    useEffect(() => {
        prefillShopsFromApi()
    }, [reset])

    return (
        <Form {...formItemLayout} className={styles.form} onSubmit={handleSubmit(onSubmit)}>
            <div className={styles.section}>
                <div className={mainStyles.subTitle}>
                    {res().generalData}
                </div>
                {shops !== [] ? (
                    <Form.Item label={res().chooseStoreLabel}>
                        <Controller
                            rules={{ required: true }}
                            control={control}
                            name="shop_id"
                            placeholder={res().chooseStoreLabel}
                            as={SelectField(shops[0], shops)}
                            defaultValue={shops[0]}
                        />
                        {errors.shop_id && <span style={errorStyle}>Shop must be selected</span>}
                    </Form.Item >
                ) : (
                        <LoadingOutlined />
                    )}
                <Form.Item label={res().formItemTitles.accountId}>
                    <Controller
                        control={control}
                        as={Input}
                        rules={{
                            minLength: 6, maxLength: 20, validate: {
                                doesNotContainCharacters: value => !RegExp(/[A-Za-z]+$/).test(value) || 'Must not contain characters'
                            }
                        }}
                        name="account_id"
                        defaultValue=""
                        placeholder={res().accountIdPlaceholder}
                    />
                    {errors.account_id && <span style={errorStyle}>Account ID must be between 6 and 30 numbers in length and cannot contain characters</span>}
                </Form.Item>
                <Form.Item label={res().formItemTitles.orderNumber}>
                    <Controller
                        control={control}
                        type="text"
                        as={Input}
                        rules={{
                            required: true, minLength: 8, maxLength: 30, validate: {
                                doesNotContainCharacters: value => !RegExp(/[A-Za-z]+$/).test(value) || 'Must not contain characters'
                            }
                        }}
                        name="invoice_id"
                        defaultValue=""
                        placeholder={res().invoicePlaceholder}
                    />
                    {errors.invoice_id && <span style={errorStyle}>Invoice ID must be between 8 and 30 characters, cannot contain characters, and cannot be empty</span>}
                </Form.Item>
                <Form.Item label={res().formItemTitles.orderSum}>
                    <Controller
                        control={control}
                        type="number"
                        as={Input}
                        rules={{ required: true, min: 100 }}
                        name="amount"
                        defaultValue="100"
                    />
                    {errors.amount && <span style={errorStyle}>Amount is required, and must be at least 100 KZT</span>}
                </Form.Item>
                <Form.Item label={res().formItemTitles.language}>
                    <Controller
                        control={control}
                        rules={{ required: true }}
                        name="language"
                        defaultValue={AlertLanguages.kazakh}
                        render={props => {
                            return (
                                <Radio.Group
                                    buttonStyle="solid"
                                    className={styles.radioGroup}
                                    value={props.value}
                                    onChange={e => props.onChange(e.target.value)}
                                >
                                    <Radio.Button key={1} value="KZ">KZ</Radio.Button>
                                    <Radio.Button key={2} value="RU">RU</Radio.Button>
                                    <Radio.Button key={3} value="EN">EN</Radio.Button>
                                </Radio.Group>
                            )
                        }}
                    />
                    {errors.language && <span style={errorStyle}>Language must be selected</span>}
                </Form.Item>
                <Form.Item label={res().formItemTitles.description}>
                    <Controller
                        name="description"
                        as={Input.TextArea}
                        control={control}
                        rules={{ required: true, maxLength: 400 }}
                    />
                    {errors.description && <span style={errorStyle}>Description must not be empty and max length must be 400 characters</span>}
                </Form.Item>

            </div >
            <div className={styles.section}>
                <div className={mainStyles.subTitle}>
                    {res().dataToSendLink}
                </div>
                <Form.Item label={res().formItemTitles.recieverEmail}>
                    <Controller
                        name="recipient_contact"
                        as={Input}
                        type="email"
                        control={control}
                        rules={{ required: true, pattern: /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/ }}
                    />
                    {errors.recipient_contact && <span style={errorStyle}>Must be a valid email</span>}
                </Form.Item>
                <Form.Item label={res().formItemTitles.recieverPhone}>
                    <Controller
                        name="recipient_contact_sms"
                        as={MaskedInput}
                        mask="(111) 111-1111"
                        type="tel"
                        control={control}
                        rules={{ required: true, minLength: 14, maxLength: 14 }}
                        prefix="+7"
                    />
                    {errors.recipient_contact_sms && <span style={errorStyle}>Must be a valid phone number</span>}
                </Form.Item>
            </div>
            <div className={styles.section}>
                <div className={mainStyles.subTitle} style={{ display: 'flex', justifyContent: 'space-between', alignItems: 'center' }}>
                    {res().paymentNotificationData}
                    <div>
                        <span style={{ marginRight: '10px' }}>Email</span>
                        <Switch defaultChecked={notifierSMSSelected} onChange={() => setNotifierSMSSelected(!notifierSMSSelected)} />
                        <span style={{ marginLeft: '10px' }}>Text</span>
                    </div>
                </div>
                {notifierSMSSelected ? (
                    <Form.Item label={res().formItemTitles.phone}>
                        <Controller
                            name="notifier_contact_sms"
                            as={MaskedInput}
                            mask="(111) 111-1111"
                            prefix="+7"
                            type="tel"
                            control={control}
                            rules={{ minLength: 14, maxLength: 14 }}
                        />
                    </Form.Item>
                ) : (
                        <Form.Item label={res().formItemTitles.email}>
                            <Controller
                                name="notifier_contact"
                                as={Input}
                                type="email"
                                control={control}
                                rules={{ pattern: /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/ }}
                            />
                        </Form.Item>
                    )}
            </div>
            <div
                className={`${layoutStyles.contentFooter} ${styles.footer}`}
            >
                <Button size="large" type="primary">
                    {sendingRequest ? (
                        <LoadingOutlined />
                    ) : (
                            <input style={{ background: 'none', border: 'none' }} type="submit" value="Submit" />
                        )}
                </Button>
                <Button size="large" onClick={onClose}>
                    {res().closeButton}
                </Button>
            </div>
        </Form >
    )
}

export default CreateInvoiceForm

I am receiving an error that states:

ERROR in C:/Users/aldiyarbat/dev/halyk-mp/node_modules/react-hook-form/dist/types/utils.d.ts(4,11):
TS2322: Type 'string | number' is not assignable to type 'string'.
  Type 'number' is not assignable to type 'string'.

I hope someone would be able to help me pin down what causes this error. Any help is greatly appreciated!

**Comment by aldibatyr** [comment]
I should point out that it runs in dev mode, but the error prevents the build.

**Reply by bluebill1049** [reply]
which version of TS r u runing?

**Reply by aldibatyr** [reply]
I am on 4.0.5

**Comment by aldibatyr** [comment]
export declare type EmptyObject = {
    [K in string | number]: never;
}; 

this is the portion of the code that triggers the error.

It is located in node_modules/react-hook-form/dist/types/utils.d.ts(4,11):

**Reply by gavranha** [reply]
Same with me, when yarn type-checks 
TS 4.0.0
works in DEV, but not BUILD

if you run 'yarn type-check' you may get something like this:

info  - Checking validity of types .jslopes@thinkpad ~/www/fxforecast.app $ yarn type-check
.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:24:77 - error TS1005: '?' expected.

24 declare type PathImpl<K extends string | number, V> = V extends Primitive ? `${K}` : `${K}` | `${K}.${Path<V>}`;
                                                                               ~~~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:24:84 - error TS1005: ';' expected.

24 declare type PathImpl<K extends string | number, V> = V extends Primitive ? `${K}` : `${K}` | `${K}.${Path<V>}`;
                                                                                      ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:24:110 - error TS1005: '(' expected.

24 declare type PathImpl<K extends string | number, V> = V extends Primitive ? `${K}` : `${K}` | `${K}.${Path<V>}`;
                                                                                                                ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:156 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                              ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:157 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                               ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:165 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                       ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:166 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                        ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:171 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                             ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:184 - error TS1005: '(' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                                          ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:190 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                                                ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:191 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                                                 ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:196 - error TS1005: ',' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                                                      ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:31:209 - error TS1005: '(' expected.

31 declare type ArrayPathImpl<K extends string | number, V> = V extends Primitive ? never : V extends ReadonlyArray<infer U> ? U extends Primitive ? never : `${K}` | `${K}.${ArrayPath<V>}` : `${K}.${ArrayPath<V>}`;
                                                                                                                                                                                                                   ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:81 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                   ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:82 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                    ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:89 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                           ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:93 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                               ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:100 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                                      ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:189 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                                                                                                                               ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:190 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                                                                                                                                ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:321 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                                                                                                                                                                                                                                                                   ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:38:322 - error TS1005: ',' expected.

38 export declare type PathValue<T, P extends Path<T> | ArrayPath<T>> = P extends `${infer K}.${infer R}` ? K extends keyof T ? R extends Path<T[K]> ? PathValue<T[K], R> : never : K extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? PathValue<V, R & Path<V>> : never : never : P extends keyof T ? T[P] : P extends `${ArrayKey}` ? T extends ReadonlyArray<infer V> ? V : never : never;
                                                                                                                                                                                                                                                                                                                                    ~

.yarn/$$virtual/react-hook-form-virtual-f177960068/0/cache/react-hook-form-npm-7.0.4-fcdaf5d875-8d2ed4522f.zip/node_modules/react-hook-form/dist/types/utils.d.ts:45:1 - error TS1160: Unterminated template literal.

45

**Reply by bluebill1049** [reply]
v7 or v6?

**Reply by gavranha** [reply]
v7

**Reply by bluebill1049** [reply]
v7, need TS 4.1 above

**Reply by gavranha** [reply]
thank you @bluebill1049

**Reply by tassiorego** [reply]
Thank you @bluebill1049

**Reply by ghasemi1994** [reply]
hi, every body.
my typescript is version 4.5 but I have error .
help me !!!!!!!!

37 |  * @typeParam T - type which may be converted to a {@link Key}
    38 |  */
  > 39 | export declare type ToKey<T> = T extends ArrayKey ? `${T}` : AsKey<T>;
       |                                                     ^
    40 | /**
    41 |  * Type which describes a path through an object
    42 |  * as a list of individual {@link Key}s.

**Comment by ghasemi1994** [comment]
my typescript version is 4.5 buy i have this error
37 |  * @typeParam T - type which may be converted to a {@link Key}
    38 |  */
  > 39 | export declare type ToKey<T> = T extends ArrayKey ? `${T}` : AsKey<T>;
       |                                                     ^
    40 | /**
    41 |  * Type which describes a path through an object
    42 |  * as a list of individual {@link Key}s.

**Comment by leapful** [comment]
@ghasemi1994 : Could you please create a new discussion for your problem ? This discussion is already expired and might be not related to the current version of react-hook-form anymore.

P.S: It would be great if your new created discussion has some codesandbox as  a sample to analyze further on the issue.
